DOCSP-35938: Connection Guide docs (#2881)
jenssegers/laravel-mongodb: docs/fundamentals/connection/connect-to-mongodb.txt

A **connection URI**, also known as a connection string, specifies how
{+odm-short+} connects to MongoDB and how to behave while connected.

Parts of a Connection URI
~~~~~~~~~~~~~~~~~~~~~~~~~

The following figure explains each part of a sample connection URI:

.. figure:: /includes/figures/connection_uri_parts.png
   :alt: Parts of a connection URI

In this connection URI, ``mongodb+srv`` is the protocol, which uses the
:manual:`DNS Seed List Connection Format </reference/connection-string/#dns-seed-list-connection-format>`
for greater flexibility in your deployment and the ability to change the
servers in rotation without reconfiguring clients.

If the machine that hosts your MongoDB deployment does not support this
feature, use protocol for the
:manual:`Standard Connection String Format </reference/connection-string/#std-label-connections-standard-connection-string-format>`
instead.

If you use a password-based authentication, the part of the connection
string after the protocol contains your username and password. Replace the
placeholder for ``user`` with your username and ``pass`` with your password.
If you use an authentication mechanism that does not require a username

> and password, omit this part of the connection URI.

Reviewer: `username` and `password` can be in the URI, or in the config:

```php
        'mongodb' => [
            'driver' => 'mongodb',
            'host' => '127.0.0.1:27017',
            'database' => 'laravel',
            'username' => 'justin',
            'password' => 'case',
        ]
```

Author: Will add a section to describe the config/database.php file

The part of the connection string after the credentials specifies your MongoDB
instance's hostname or IP address and port. The preceding example uses
``sample.host`` as the hostname and ``27017`` as the port. Replace these values
to point to your MongoDB instance.

The last part of the connection string specifies connection and authentication
options. In the example, we set the following connection options and values:

- ``maxPoolSize=20``
- ``w=majority``

> 

Author: Note:
This section will include a link to the [Connection Options](https://jira.mongodb.org/browse/DOCSP-35941) page once completed.

   DB_URI="mongodb://127.0.0.1:27017/";

To learn how to download  and install MongoDB server, see
:manual:`Install MongoDB Community Edition </installation/#mongodb-installation-tutorials>`
in the {+server-docs-name+}.

.. _laravel-connect-replica-set:

Connect to a Replica Set
~~~~~~~~~~~~~~~~~~~~~~~~

A MongoDB replica set deployment is a group of connected instances, or nodes,
where the nodes store the same data set. This configuration of instances
provides data redundancy and high data availability.

To connect to a replica set deployment, specify each node's hostname and port
number, separated by commas, and the replica set name as the value of the
``replicaSet`` parameter in the connection string.

This example, which shows the connection string you can add to your
Laravel application's ``.env`` file to connect to a replica set, uses the
following sample values:

- ``host1``, ``host2``, and ``host3`` as the hostnames of the MongoDB nodes
- ``27017`` as the port on which MongoDB runs on those hosts
- ``myRS`` as the configured name of the replica set
- ``myUser`` and ``myPass123`` as the credentials of a database user

.. code-block:: bash

>    DB_URI="mongodb://myUser:myPass123@host1:27017,host2:27017,host3:27017/?replicaSet=myRS"

Reviewer: After testing on a local replica set, I can confirm that the `replicatSet` option works.
```env
MONGODB_URI="mongodb://localhost:27017,localhost:27018,localhost:27019/"
```
```php
        'mongodb' => [
            'driver' => 'mongodb',
            'dsn' => env('MONGODB_URI'),
            'database' => 'laravel',
            'options' => [
                'replicaSet' => 'myRS',
            ]
        ]
```

You can also pass the list of hosts like this:
```php
        'mongodb' => [
            'driver' => 'mongodb',
            'host' => [
                '127.0.0.1:27017',
                '127.0.0.1:27018',
                '127.0.0.1:27019',
            ],
            'database' => 'laravel',
            'options' => [
                'replicaSet' => 'myRS',
            ]
        ]
```

instead.

If you use a password-based authentication, the part of the connection
string after the protocol contains your username and password. Replace the
placeholder for ``user`` with your username and ``pass`` with your password.
If you use an authentication mechanism that does not require a username
and password, omit this part of the connection URI.

The part of the connection string after the credentials specifies your MongoDB
instance's hostname or IP address and port. The preceding example uses
``sample.host`` as the hostname and ``27017`` as the port. Replace these values
to point to your MongoDB instance.

The last part of the connection string specifies connection and authentication
options. In the example, we set the following connection options and values:

- ``maxPoolSize=20``
- ``w=majority``

.. _laravel-database-config:

Laravel Database Connection Configuration
------------------------------------------

{+odm-short+} lets you configure your MongoDB database connection in the
``config/database.php`` Laravel application file. You can specify the following
connection details in this file:

> - ``default``, which specifies the default database connection to use when
>   unspecified

Reviewer 2: ```suggestion
- ``default``, which specifies the default database connection to use
```

Author: Thanks for the suggestion. I wanted to emphasize the behavior of the default value, but now think maybe including "default" in the description of itself might be confusing. think maybe a more clear way to describe this could be:
"default, which specifies the database connection to use when unspecified"

> - ``connections``, which contains database connection information to access
>   various databases from your application

Reviewer 2: ```suggestion
- ``connections``, which contains database connection information to access
  one or more databases from your application
```

You can use the following code in the configuration file to set the default
connection to a corresponding ``"mongodb"`` entry in the ``connections`` array:

.. code-block:: php

>    'default' => 'mongodb',

Reviewer 2: Q: is the comma intentional?

Author: Thanks for asking. Yes, since it's the convention that the PHP DBX team requested (commas after each item of a list) as this is part of an (associative) array of items.

For a MongoDB database connection, you can specify the following details:

- ``driver``, which specifies the database driver to use for the connection
- ``dsn``, the data source name (DSN) that specifies the MongoDB connection URI
- ``host``, which you can use instead of the ``dsn`` setting to specify the
  network address of one or more MongoDB nodes
- ``database``, which specifies the name of the MongoDB database to read and
  write to
- ``username`` and ``password``, which you can optionally include to specify

>   your database user's credentials to authenticate with MongoDB

Reviewer 2: I: right now, it's not very clear that using different properties such as host username password port, etc. is an alternative to including all of this information in the DSN

S: add an admonition or revise these descriptions to make this clear

Author: Good idea!  Will add an admonition to explain.

- ``options`` and ``driverOptions``, which specify connection options to

>   pass to the MongoDB driver

Reviewer 2: S: maybe converting this into a table could be helpful

>   pass to the MongoDB driver

Reviewer 2: Q: is there a place to find a complete list of what goes under each property? Or will this be covered in the connection options page

Author: Thanks for asking. These are covered in the Connection Options ticket and a link will be added to this page per https://github.com/mongodb/laravel-mongodb/pull/2881#discussion_r1571464140

   :caption: Sample config/database.php connection entry
   :emphasize-lines: 3

   'connections' => [
       'mongodb' => [
           'dsn' => env('DB_URI'), // uses the value of the DB_URI environment variable
           'driver' => 'mongodb',
           'database' => 'sample_mflix',
           // ...
       ],
     // ...
   ]

.. tip::

   To retrieve your Atlas connection string, follow the
   :ref:`Create a Connection String <laravel-quick-start-connection-string>`
   step of the Quick Start tutorial.

.. _laravel-other-ways-to-connect:

Other Ways to Connect to MongoDB
--------------------------------

The following sections show you how to connect to a single MongoDB server
instance or a replica set not hosted on MongoDB Atlas.

.. _laravel-connect-localhost:

Connect to a MongoDB Server on Your Local Machine
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This section shows an example connection string you can use when running a
Laravel application and MongoDB server from the same machine, such as your
local development environment.

To connect your application to a MongoDB instance hosted on the same machine,
you must complete the following tasks:

> - Download, install, and run MongoDB server.

Reviewer 2: ```suggestion
- Download, install, and run the MongoDB server.
```

If you use an authentication mechanism that does not require a username
and password, omit this part of the connection URI.

The part of the connection string after the credentials specifies your MongoDB
instance's hostname or IP address and port. The preceding example uses
``sample.host`` as the hostname and ``27017`` as the port. Replace these values
to point to your MongoDB instance.

The last part of the connection string specifies connection and authentication
options. In the example, we set the following connection options and values:

- ``maxPoolSize=20``
- ``w=majority``

.. _laravel-database-config:

Laravel Database Connection Configuration
------------------------------------------

{+odm-short+} lets you configure your MongoDB database connection in the
``config/database.php`` Laravel application file. You can specify the following
connection details in this file:

- ``default``, which specifies the database connection to use when unspecified
- ``connections``, which contains database connection information to access
  one or more databases from your application

You can use the following code in the configuration file to set the default

> connection to a corresponding ``"mongodb"`` entry in the ``connections`` array:

Reviewer 2: ```suggestion
connection to a corresponding ``mongodb`` entry in the ``connections`` array:
```

.. code-block:: php

   'default' => 'mongodb',

For a MongoDB database connection, you can specify the following details:

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Setting
     - Description

   * - ``driver``
     - Specifies the database driver to use for the connection.

   * - ``dsn``
     - The data source name (DSN) that specifies the MongoDB connection URI.

   * - ``host``
     - Specifies the network address of one or more MongoDB nodes in a

>        deployment. You can use this setting instead of the ``dsn`` setting.

Reviewer: This is `host:port` or a list of `host:port`.

Examples:
- `'host' => 'localhost:27017'`
- `'host' => ['server1.example.net:27017', 'server2.example.net:27017', 'server3.example.net:27017']`

Author: Question: from my experimentation, this option only accepts non-seedlist hostnames (``mongodb`` and not ``mongodb+srv``), correct?

Reviewer: In fact, we need to add an option for `mongodb+srv`.
https://github.com/mongodb/laravel-mongodb/blob/d0978a80f92fb3ea0f752f86994e883ca46c7544/src/Connection.php#L255

   * - ``database``
     - Specifies the name of the MongoDB database to read and write to.

   * - ``username``
     - Specifies your database user's username credential to authenticate
       with MongoDB.

   * - ``password``
     - Specifies your database user's password credential to authenticate
       with MongoDB.

   * - ``options``
     - Specifies connection options to pass to MongoDB that determine the
       connection behavior.

   * - ``driverOptions``
     - Specifies options specific to pass to the MongoDB PHP Library driver

>        that determine the driver behavior for that connection.

Reviewer: You should link to the doc of options: https://www.mongodb.com/docs/php-library/current/reference/method/MongoDBClient__construct/#parameters

Author: Thanks, I will add links from options and driverOptions to the Connection Options page (which includes the link you provided) once merged as a [cleanup task](https://jira.mongodb.org/browse/DOCSP-38344).

               'maxPoolSize' => 20,
               'w' => 'majority',
           ],
           'driverOptions' => [
               'serverApi' => 1,
           ],
       ],
       // ...
   ],

The following sections provide common ways of specifying MongoDB connections.

.. _laravel-atlas-connection-example:

Connection Example
------------------

This section shows how to configure your Laravel application's DSN by using a
MongoDB Atlas connection string.

To add your MongoDB DSN to your Laravel application, make the following changes:

- Add the DSN as an environment variable in your project's ``.env`` environment
  configuration file. Set the variable value to your Atlas connection string.
- Add a connection entry for your MongoDB connection in the ``connections``
  array of your ``config/database.php`` configuration file. Set the ``dsn``
  value of the connection entry to reference the environment variable that
  contains your DSN.

The following examples show how to specify ``"mongodb+srv://myUser:myPass123@mongodb0.example.com/"``
as the connection string in the relevant configuration files:

.. code-block:: php
   :caption: Sample .env environment configuration

>    DB_URI=""mongodb+srv://myUser:myPass123@mongodb0.example.com/";

Reviewer 2: ```suggestion
   DB_URI="mongodb+srv://myUser:myPass123@mongodb0.example.com/"
```

Author: Whoops, great catch!
I should also use "bash" instead of "php" highlighting.
